### ISSUE TITLE
Questions about the implementation.

### ISSUE THREAD

**Issue by kleysonr** [AdrianUng/keras-triplet-loss-mnist#1]
Hi,

Thanks for the code. I've being trying to use the tf.contrib.losses.metric_learning.triplet_semihard_loss with a siamese network but struggling with that.

May you help me clarifying some questions ?

1 - Why did you changed the triplet_semihard_loss function ?

2 - In the blogpost https://omoindrot.github.io/triplet-loss and AndrewNG's lecture it's supposed we have a siamese network with 3 inputs to have the triplet loss calculated. In your implementation you just have a single input. Why did you do that in this way ?

3 - Is not possible to train without having the labels as an input in the network ?

Best Regards.
Kleyson Rios.

**Comment by AdrianUng** [comment]
Hello, Kleyson

Thank you for your questons! Sorry for the late reply, I just noticed the issue...
Here are my answers:

Q1. I actually didn't change the function. There are only 3 lines that were commented (at the beginning of the function), mainly dealing with asserting a certain shape for the data/labels. The rest of the script is left untouched. If you want you could easily call the tf-function within your locally declared keras-loss.

Q2. You are correct about the 3 inputs, but there are several reasons to consider this implementation. First of all, if you have 3 rigid inputs, it means you need 3 networks (each with its own set of weights) which output an embedding corresponding to the input. This is very costly in terms of memory (fitting 3 nets) as it is in terms of time (generating the pairs so that all nets learn the same thing). By passing the 3 inputs to the same network you basically end up with a scenario where the '3 branches' share weights. Please check out the keras-team implementation of the siamese network [link](https://github.com/keras-team/keras/blob/master/examples/mnist_siamese.py). The triplet loss requires more flexibility in terms of how the pairs are constructed (generating the triplets at batch-level), which is not necessarily  the case with the Contrastive Loss.

Q3. There probably is a way to do that, but my implementation represents a quick solution based on Keras's **model.fit()** function. I would recommend looking into data_generators in Keras as a candidate for that implementation. Check out this [link](https://www.pyimagesearch.com/2018/12/24/how-to-use-keras-fit-and-fit_generator-a-hands-on-tutorial/). I plan to include such a solution at some point in the near future.

I hope these answers help! Let me know if you have more questions :)

Best Regards,
Adrian

**Comment by kleysonr** [comment]
Thanks Adrian.
